staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1366,28 +1324,76 @@ func hasServedCRDVersion(spec *apiextensionsv1.CustomResourceDefinitionSpec, ver
 	return false
 }
 
+func structuralSchemasFromCRD(crd *apiextensionsv1.CustomResourceDefinition) (map[string]*structuralschema.Structural, error) {
+	structuralSchemas := map[string]*structuralschema.Structural{}
+	for _, v := range crd.Spec.Versions {
+		val := v.Schema
+		if val == nil {
+			continue
+		}
+		internalValidation := &apiextensionsinternal.CustomResourceValidation{}
+		if err := apiextensionsv1.Convert_v1_CustomResourceValidation_To_apiextensions_CustomResourceValidation(val, internalValidation, nil); err != nil {
+			return nil, fmt.Errorf("failed converting CRD validation to internal version: %v", err)
+		}
+
+		if validation.SchemaHasInvalidTypes(internalValidation.OpenAPIV3Schema) {
+			return nil, fmt.Errorf("schema for CRD %v has invalid types", crd.Name)
+		}
+
+		s, err := structuralschema.NewStructural(internalValidation.OpenAPIV3Schema)
+		if !crd.Spec.PreserveUnknownFields && err != nil {
+			// This should never happen. If it does, it is a programming error.
+			utilruntime.HandleError(fmt.Errorf("failed to convert schema to structural: %v", err))
+			return nil, fmt.Errorf("the server could not properly serve the CR schema") // validation should avoid this
+		}
+
+		if !crd.Spec.PreserveUnknownFields {
+			// we don't own s completely, e.g. defaults are not deep-copied. So better make a copy here.
+			s = s.DeepCopy()
+
+			if err := structuraldefaulting.PruneDefaults(s); err != nil {
+				// This should never happen. If it does, it is a programming error.
+				utilruntime.HandleError(fmt.Errorf("failed to prune defaults: %v", err))
+				return nil, fmt.Errorf("the server could not properly serve the CR schema") // validation should avoid this
+			}
+		}
+		structuralSchemas[v.Name] = s
+	}
+
+	return structuralSchemas, nil
+}
+
 // buildOpenAPIModelsForApply constructs openapi models from any validation schemas specified in the custom resource,
 // and merges it with the models defined in the static OpenAPI spec.
 // Returns nil models ifthe static spec is nil, or an error is encountered.
-func buildOpenAPIModelsForApply(staticOpenAPISpec *spec.Swagger, crd *apiextensionsv1.CustomResourceDefinition) (*spec.Swagger, error) {
+func buildOpenAPIModelsForApply(staticOpenAPISpec *spec.Swagger, crd *apiextensionsv1.CustomResourceDefinition) (fieldmanager.TypeConverter, map[string]*structuralschema.Structural, error) {
```

Comment:
Not a big fan of the signature, why would `buildOpenAPIModelsForApply` also build the structural schemas?

Reply:
simplified to only return type converter, and take structural schemas as input